internal/controller/controllers/agentlabel_controller.go

```diff
@@ -0,0 +1,191 @@
+	log := r.Log.WithFields(
+		logrus.Fields{
+			"agent_label":           req.Name,
+			"agent_label_namespace": req.Namespace,
+		})
+
+	defer func() {
+		log.Info("AgentLabel Reconcile ended")
+	}()
+
+	log.Info("AgentLabel Reconcile started")
+
+	agent := &aiv1beta1.Agent{}
+	if err := r.Get(ctx, req.NamespacedName, agent); err != nil {
+		log.WithError(err).Errorf("Failed to get AgentClassification %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Get the inventory into interfaces by way of json marshal/unmarshal
+	var inventoryInterface interface{}
+	jsonInventory, _ := json.Marshal(agent.Status.Inventory)
+	_ = json.Unmarshal(jsonInventory, &inventoryInterface)
+
+	classifications := aiv1beta1.AgentClassificationList{}
+	opts := &client.ListOptions{
+		Namespace: agent.Namespace,
+	}
+	err := r.List(ctx, &classifications, opts)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	changed := false
+	for _, classification := range classifications.Items {
+		if !classification.DeletionTimestamp.IsZero() {
+			log.Infof("classification %s is being deleted", classification.Name)
+			changed = deleteAgentLabel(log, agent, ClassificationLabelPrefix+classification.Spec.LabelKey, classification.Spec.LabelValue) || changed
+			continue
+		}
+
+		query, err := gojq.Parse(classification.Spec.Query)
+		if err != nil {
+			// Should not happen - validated via webhook
+			log.Errorf("Failed to parse query: %s\n", query)
+			changed = setAgentLabel(log, agent, ClassificationLabelPrefix+classification.Spec.LabelKey, queryErrorValue(classification.Spec.LabelValue)) || changed
+			continue
+		}
+
+		matched, err := checkMatch(log, query, inventoryInterface)
+		if err != nil {
+			changed = setAgentLabel(log, agent, ClassificationLabelPrefix+classification.Spec.LabelKey, queryErrorValue(classification.Spec.LabelValue)) || changed
+		} else if !matched {
+			changed = deleteAgentLabel(log, agent, ClassificationLabelPrefix+classification.Spec.LabelKey, classification.Spec.LabelValue) || changed
+		} else {
+			changed = setAgentLabel(log, agent, ClassificationLabelPrefix+classification.Spec.LabelKey, classification.Spec.LabelValue) || changed
+		}
+	}
+
+	if changed {
+		if err := r.Update(ctx, agent); err != nil {
+			log.WithError(err).Error("failed to update agent")
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func queryErrorValue(originalValue string) string {
+	return fmt.Sprintf("QUERYERROR-%s", originalValue)
+}
+
+func checkMatch(log *logrus.Entry, query *gojq.Query, inventoryInterface interface{}) (bool, error) {
+	iter := query.Run(inventoryInterface)
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			return false, nil
+		}
+		if err, ok := v.(error); ok {
+			return false, err
+		}
+		if res, ok := v.(bool); ok {
+			if res {
+				return true, nil
+			}
+		}
+	}
```

Comment:
```suggestion
	values := []interface{}{}
	for {
		v, ok := iter.Next()
		if !ok {
			break;
		}

		if err, ok := v.(error); ok {
			return false, err
		}

		values = append(values, v)
	}

	if len(values) == 0 {
		return false, errors.New("Expected boolean, found no values")
	}

	if len(values) > 1 {
		return false, errors.New("Expected boolean, found multiple values")
	}

	value := values[0]

	if res, ok := value.(bool); ok {
		if res {
			return true, nil
		}
	}

	return false, nil
```

Do not accept `echo '[true, false, false]' | jq '.[]'`

Reply:
Fixed, thank you!